README
This is a fork of ATS/Anairiats which contains some experimental
support of Unicode.

changes.txt
########################################################################

ats-lang-anairiats-0.2.10:

1. arrsz -> arrpsz
   arraysize -> arrayptrsize
2. list0_of_list1 -> list0_of_list
   list1_of_list0 -> list_of_list0
3. Fixing a glitch involving tail-call optimization
4. Fixing a glitch involving exceptions declared in prelude/SATS
5. Fixing a minor flaw in atsopt-depgen
6. Adding linset_copy in libats/linset_avltree

########################################################################

ats-lang-anairiats-0.2.9:

1. Fixing some macros in ccomp/runtime/ats_exception.h
2. Removing the field tmpvar_root in tmpvar and related functions.
3. Adding support for using CMake to build projects written in ATS.
   Kudos to Hanwen!
4. $showtype support for debugging is added (2012-11-24)
5. Adding libats/linmap_skiplist for linear maps based on skip-lists.
6. ((<exp1>, ..., <exp2>)) -> @(<exp1>, ..., <exp2>)
7. If the assigned external name of a function is empty, then its
   internal name should be used in the place where its external name
   is needed. For example, the following declarations are equivalent:
   fun foo (x: int): void = "ext#"
   fun foo (x: int): void = "ext#foo"
   The following two are equivalent as well:
   fun bar (x: int): void = "mac#"
   fun bar (x: int): void = "mac#bar"

########################################################################

ats-lang-anairiats-0.2.8:

1. Fixing a bug involving assumed abstract types.
2. Adding list_mapopt_* functions.
3. Packaging libatsdoc into ATS-0.2.8
4. Packaging utils/atsdoc into ATS-0.2.8
5. Adding contrib/mysql
6. Adding contrib/jansson (based on previous work by Chris Double)

########################################################################

ats-lang-anairiats-0.2.7:

1. free_gc_viewt0ype_addr_view -> free_gc_t0ype_addr_view
   free_gc_viewt0ype_int_addr_view -> free_gc_t0ype_int_addr_view
2. ptr_free and array_ptr_free: interface change
3. Adding funheap_binomial and linheap_binomial
4. Adding libats/SATS/bimatrix.sats. Thanks to Artyom!
5. The s2exp_syneq function (syn. eq.) now employs eta-expansion
6. !READ(string) -> string (SHARED is introduced to do the opposite)
7. Adding libats/lockptr_spin and libats/lockref_spin (spinlocks)
8. The docbook source for the ATS intro book is now included.
9. The closures in each infile object are made to be linear and they
   are freed when the infile object is freed.
10. The compiler now generates code to prevent an abstract type from
    being implemented (assumed) more than once.
11. && and || are now macros (long overdue!)
12. The libats/ngc/linmap_pavltree package is added. Kudos to Artyom!
13: Adding libats/funmset_listord (functional multiset implementation)
14: Fixing a bug involving tailcall optimization (argument movement)
15: Fixing a minor bug involving p2at_arg_tr.
16: The interface for pthread creation in ATS is slightly modified.

########################################################################

ats-lang-anairiats-0.2.6:

1. Adding contrib/JNI for interface ATS and Java.
2. Adding the --xrefprelude flag to make cross-referencing work again.
3. The keyword [sta] can now also be written as [stacst].
4. If a name refers to several static constants, then the latest declared
   non-functional one is chosen (if there is one).
5. The tutorial (doc/TUTORIAL) is no longer put into the release package.
6. Allowing #FILENAME and #LOCATION to be used as e0xp. For instance, the
   following line is now legal:
   #print (#FILENAME)
7. Fixing a bug related to xref that caused illegal xhtml files to be
   generated (cases of A elements being embedded inside another A element).
   Thanks, David!
8. Introducing ISALIGNED to handle alignment.
9. Introducing NULLABLE to handle ptr_zero and ptr_zero_tsz.
10. Adding libats/funmap_rbtree (functional maps based on red-black trees)
11. Adding contrib/gtkglext (based on gtkglext-1.0)
12. Adding contrib/atspslide for doing slide presentation and other things.
13: Adding libc/gdbm/gdbm (API for gdbm in ATS).
14: Adding libc/gdbm/ndbm (API for gdbm-ndbm in ATS).
15: Adding libats/ngc/DATS/dlist.dats. Thanks to Artyom!
16: [atslib] can now process flags for each individual file. See .libfiles.
17: pervasive changes: free_gc_v (T, ...) => free_gc_v (T?, ...)

########################################################################

ats-lang-anairiats-0.2.5:

1. Fixing a bug in handling call-by-reference. I was really surprised that
   the bug had not shown up earlier.
2. input_line_vt is added in prelude/filebas for returning a linear string
3. Fixing a bug in [valprim_is_mutable]: casting was incorrectly handled.
4. atslib_feof and atslib_ferror are changed to inlined functions (instead
   of macros). This is done to fix Cgwin's defining feof as a macro.
5. Some changes are made to support ATS on FreeBSD.
6. Fixing a bug involving linear closures (of the form <lin,cloptr>).
7. Adding support for if!: if! (e_cond, e_then, e_else) is for the
   following expression: (if e_cond then e_then else e_else).
8. Adding libats/funset_avltree to support (functional) sets.
9. The original *_clo and *_cloptr functions are renamed as *_vclo and
   *_vcloptr respectively, and the new *_clo and *_cloptr functions no
   longer require a (linear) proof argument.
10. [gc_stack_get_dir] is re-written to fix a bug triggered due to a form
    of over-optimization by gcc-4.6: a volatile function may get inlined.
    Kudos to Matthew for helping me on this one!
11. Fixing a bug that wrongly affected the exn-effect tracking.
12. Adding contrib/GLES2 to support OpenGL/ES2 binding. Thanks to Artyom!
13. Renaming: list_vt_*_clo -> list_vt_*_vclo. Also adding various functions:
    list_vt_*_cloptr and list_vt_*_vcloptr
14. Makefile_macosx is added for building ATS on MAC OS X. Thanks to Georg!
15. Renaming (in various packages): *_clo -> *_vclo and *_cloptr -> *_vcloptr
16. Adding libatsdoc.
17. Adding the [atsdoc] utility for writing documents

ccomp/runtime/ats_prelude.c
/************************************************************************/
/*                                                                      */
/*                         Applied Type System                          */
/*                                                                      */
/*                              Hongwei Xi                              */
/*                                                                      */
/************************************************************************/

/*
** ATS - Unleashing the Power of Types!
**
** Copyright (C) 2002-2008 Hongwei Xi.
**
** ATS is free software;  you can  redistribute it and/or modify it under
** the terms of the GNU LESSER GENERAL PUBLIC LICENSE as published by the
** Free Software Foundation; either version 2.1, or (at your option)  any
** later version.
** 
** ATS is distributed in the hope that it will be useful, but WITHOUT ANY
** WARRANTY; without  even  the  implied  warranty  of MERCHANTABILITY or
** FITNESS FOR A PARTICULAR PURPOSE.  See the  GNU General Public License
** for more details.
** 
** You  should  have  received  a  copy of the GNU General Public License
** along  with  ATS;  see the  file COPYING.  If not, please write to the
** Free Software Foundation,  51 Franklin Street, Fifth Floor, Boston, MA
** 02110-1301, USA.
*/

/* ****** ****** */

/* author: Hongwei Xi (hwxi AT cs DOT bu DOT edu) */

/* ****** ****** */

#include "config.h"

#include <stdio.h>

#ifdef _ATS_MULTITHREAD
#include <pthread.h>
#endif // end of [_ATS_MULTITHREAD]

/* ****** ****** */

#include "ats_types.h"
#include "ats_basics.h"
#include "ats_memory.h"

/* ****** ****** */
//
// HX: implemented in [prelude/DATS/basics.dats]
//
extern void ats_exit_errmsg (int err, char *msg) ;

/* ****** ****** */

/*
** sizeof(ats_empty_type) == 0
*/
ats_empty_type ats_empty_value ;

/* ****** ****** */
//
// HX: The following variables are used in basics.dats
//
int ats_stdin_view_lock = 1 ;
int ats_stdout_view_lock = 1 ;
int ats_stderr_view_lock = 1 ;

/* ****** ****** */

/*
** the type of [the_ats_exception_stack]
** is given in the file [ats_exception.h]
*/

#ifdef _ATS_MULTITHREAD
ATSthreadlocalstorage() // thread-local storage
#endif // end of [_ATS_MULTITHREAD]
void *the_ats_exception_stack = NULL ;

/* ****** ****** */
//
// HX: some common exceptions
//
ats_exn_type
AssertionExceptionCon = { 10, "AssertionException" } ;
ats_exn_ptr_type AssertionException = &AssertionExceptionCon ;

ats_exn_type
OverflowExceptionCon = { 20, "OverflowException" } ;
ats_exn_ptr_type OverflowException = &OverflowExceptionCon ;

ats_exn_type
DivisionByZeroExceptionCon = { 30, "DivisionByZeroException" } ;
ats_exn_ptr_type DivisionByZeroException = &DivisionByZeroExceptionCon ;

/* ****** ****** */

ats_exn_type
ListSubscriptExceptionCon = { 40, "ListSubscriptException" } ;
ats_exn_ptr_type ListSubscriptException = &ListSubscriptExceptionCon ;
ats_exn_ptr_type
ListSubscriptException_make () { return ListSubscriptException ; }
ats_bool_type
isListSubscriptException (ats_exn_ptr_type exn) {
  return (ListSubscriptException->tag==exn->tag ? ats_true_bool : ats_false_bool) ;
}

ats_exn_type
ArraySubscriptExceptionCon = { 50, "ArraySubscriptException" } ;
ats_exn_ptr_type ArraySubscriptException = &ArraySubscriptExceptionCon ;
ats_exn_ptr_type
ArraySubscriptException_make () { return ArraySubscriptException ; }
ats_bool_type
isArraySubscriptException (ats_exn_ptr_type exn) {
  return (ArraySubscriptException->tag==exn->tag ? ats_true_bool : ats_false_bool) ;
}

/* ****** ****** */
//
// HX: the numbers less than 1024 are all
int ats_exception_con_tag = 1024 ; // reserved for special use

/* ****** ****** */

/*
** function for handling uncaught exceptions
*/

extern void exit (int status) ; // declared in [stdlib.h]

ats_void_type
ats_uncaught_exception_handle (
  const ats_exn_ptr_type exn
) {
  fprintf(stderr,
    "exit(ATS): uncaught exception: %s(%d)\n", exn->name, exn->tag
  ) ; exit(1) ;
  return ; // deadcode
} // end of [ats_uncaught_exception_handle]

/* ****** ****** */

/*
** functions for handling match failures
*/

void
ats_caseof_failure_handle (
  const char *loc // location of the failure
) {
  fprintf(stderr, "exit(ATS): %s: match failure.\n", loc) ; exit(1) ;
  return ; // deadcode
} // end of [ats_caseof_failure_handle]

void
ats_funarg_match_failure_handle (
  const char *loc // location of the failure
) {
  fprintf(stderr, "exit(ATS): %s: funarg match failure.\n", loc) ; exit(1) ;
  return ; // deadcode
} // end of [ats_funarg_match_failure_handle]

/* ****** ****** */

/*
** functions for memory allocation and deallocation
*/

#ifdef _ATS_NGC // no GC
#include "ats_prelude_ngc.c"
#elif _ATS_GCATS // special GC for ATS
#include "ats_prelude_gcats.c"
#elif _ATS_GCATS2 // special GC for ATS
#include "ats_prelude_gcats2.c"
#elif _ATS_GCBDW // Boehm-Demers-Weise conservative GC for C/C++
#include "ats_prelude_gcbdw.c"
#else // _ATS_NGC is the default
#include "ats_prelude_ngc.c"
#endif // end of [ifdef]

/* ****** ****** */

#ifdef _ATS_MULTITHREAD

/* ****** ****** */

ats_int_type
ats_pthread_create_detached (
  ats_ptr_type f, ats_ptr_type env, ats_ref_type tid_r
) {
  int ret ;
#ifdef _ATS_NGC
  pthread_attr_t attr;
  pthread_attr_init (&attr);
  pthread_attr_setdetachstate (&attr, PTHREAD_CREATE_DETACHED) ;
  ret = pthread_create ((pthread_t*)tid_r, &attr, f, env) ;
#elif _ATS_GCATS
  ret = gc_pthread_create (1/*detached*/, f, env, tid_r) ;
/*
  fprintf (stderr, "exit(ATS): there is no support for pthreads under GCATS.\n") ;
  exit (1) ;
*/
#elif _ATS_GCBDW
  fprintf (stderr, "exit(ATS): there is no support for pthreads under GCBDW.\n") ;
  exit (1) ;
#else // _ATS_NGC is the default
  pthread_attr_t attr;
  pthread_attr_init (&attr);
  pthread_attr_setdetachstate (&attr, PTHREAD_CREATE_DETACHED) ;
  ret = pthread_create ((pthread_t*)tid_r, &attr, f, env) ;
#endif // end of [#ifdef]
  return ret ;
} // end of [ats_pthread_create_detached]

/* ****** ****** */

ats_void_type
ats_pthread_exit () {
  pthread_exit (NULL) ; // HX: this may cause memory leak if not used properly
} // end of [ats_pthread_exit]

/* ****** ****** */

#endif /* [ifdef _ATS_MULTITHREAD] */

/* ****** ****** */

/* end of [ats_prelude.c] */


utils/latex/lstatslang.sty.sh
# Generates lstatslang.sty which is printed to standard output.
#
# Requires keywords.txt generated by the Makefile from ats_keywords.dats

is_directive() {
  case "$1" in
    \#FILENAME|\#LOCATION|\#CHARCOUNT|\#LINECOUNT) return 1;;  # false
    \#*) return 0;;  # true
    dynload|staload|\$dynload) return 0;;  # true
  esac

  return 1  # false
}

# Usage: cat keywords.txt | typeset 'initialkeyword'
#
typeset() {
  local -r indent=5
  local n=$indent       # line width
  local c=0             # word count

  if [[ $1 ]]; then
    ((n += ${#1}))
  fi

  while read word; do
    ((!c++)) || echo -ne ','
    if ((${#word} + n >= 70)); then
      echo -ne '%\n     '
      n=$indent
    fi
    echo -ne "$word"
    ((n += ${#word} + 1))
  done
}

quote() {
  echo "${1/\#/\\\#}"
}

keywords() {
  while read word; do
    is_directive "$word" || quote "$word"
  done < keywords.txt
}

directives() {
  while read word; do
    ! is_directive "$word" || quote "$word"
  done < keywords.txt
}

literate() {
  echo "{$1}{\$$2$ }$3"
}

literate_tight() {
  echo "{$1}{\$$2$}$3"
}

cat <<END
%%
%% This file is distributed under the terms of the LaTeX Project Public
%% License from CTAN archives in directory  macros/latex/base/lppl.txt.
%% Either version 1.3 or, at your option, any later version.
%%
%% This file is completely free and comes without any warranty.
%%
\ProvidesFile{lstatslang.sty}
    [2012/10/23 1.0 listings language file for ATS]
\newif\if@lstatslang@literate@ \@lstatslang@literate@false
\DeclareOption{literate}{\@lstatslang@literate@true}
\ProcessOptions\relax
\if@lstatslang@literate@
\RequirePackage{amssymb}
\fi
%%
%% ATS language definition (c) 2012  Matthew Danish and Likai Liu
%%
%% ATS is an ML-like dependent-typed functional language with theorem
%% proving using classical and linear propositions.
%%
\lstdefinelanguage{ATS}%
  {morekeywords={$(keywords | typeset 'morekeywords')},%
   moredirectives={$(directives | typeset 'moredirectives')},%
   sensitive=true,%
   morecomment=[l]{//},%
   morecomment=[n]{(*}{*)},%
   morestring=[b]",%
  }[keywords,comments,directives,strings]%
\lstdefinelanguage[literate]{ATS}[]{ATS}%
  {literate=
   $(literate '=' '=' 1)
   $(literate '->' '\to' 1)
   $(literate '<=' '\leq' 1)
   $(literate '<' '<' 1)
   $(literate '>=' '\geq' 1)
   $(literate '>' '>' 1)
   $(literate '<>' '\neq' 1)
   $(literate ':<>' ':_{\textrm{pure}}' 4)
   $(literate ':<prf>' ':_{\textrm{prf}}' 4)
   $(literate '-<lin,prf>' '\mathop{\multimap}_{\textrm{prf}}' 4)
   $(literate '|' '\mid' 1)
   $(literate_tight '\{' '\{' 1)
   $(literate_tight '\}' '\}' 1)
  }%
END


ccomp/runtime/ats_exception.h
/************************************************************************/
/*                                                                      */
/*                         Applied Type System                          */
/*                                                                      */
/*                              Hongwei Xi                              */
/*                                                                      */
/************************************************************************/

/*
** ATS - Unleashing the Power of Types!
**
** Copyright (C) 2002-2008 Hongwei Xi.
**
** ATS is free software;  you can  redistribute it and/or modify it under
** the terms of the GNU LESSER GENERAL PUBLIC LICENSE as published by the
** Free Software Foundation; either version 2.1, or (at your option)  any
** later version.
** 
** ATS is distributed in the hope that it will be useful, but WITHOUT ANY
** WARRANTY; without  even  the  implied  warranty  of MERCHANTABILITY or
** FITNESS FOR A PARTICULAR PURPOSE.  See the  GNU General Public License
** for more details.
** 
** You  should  have  received  a  copy of the GNU General Public License
** along  with  ATS;  see the  file COPYING.  If not, please write to the
** Free Software Foundation,  51 Franklin Street, Fifth Floor, Boston, MA
** 02110-1301, USA.
**
*/

/*
**
** Authors: 
** Likai Liu (liulk AT cs DOT bu DOT edu) // Summer 2005
** Rick Lavoie (coldfury AT cs DOT bu DOT edu) // Fall 2006
** Hongwei Xi (hwxi AT cs DOT bu DOT edu) // Summer 2007
**
*/

#ifndef ATS_EXCEPTION_H
#define ATS_EXCEPTION_H // the file should only be loaded once

/* ****** ****** */

/*
** always loaded after the following two files:
*/
#include "ats_types.h"
#include "ats_basics.h"

/* ****** ****** */

#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE // for [sigsetjmp] in gcc-4.5
#endif // end of [_XOPEN_SOURCE]
#include <setjmp.h>

/* ****** ****** */
//
// HX-2011-04-24: the function [alloca] is declared in
extern void *alloca (size_t nsz) ; // [alloca.h] or [stdlib.h]

/* ****** ****** */

/*
** function for handling uncaught exceptions
*/

extern
ats_void_type
ats_uncaught_exception_handle(const ats_exn_ptr_type exn) ;

/* ****** ****** */

/* exception implementation */

typedef
struct ats_exception_frame_struct {
  ats_exn_ptr_type exn ;
  struct ats_exception_frame_struct *prev ;
  sigjmp_buf env ;
} ats_exception_frame_type ;

/* ****** ****** */

#ifdef _ATS_MULTITHREAD
extern ATSthreadlocalstorage()
void *the_ats_exception_stack ;
#else /* single thread */
extern
void *the_ats_exception_stack ;
#endif // end of [_ATS_MULTITHREAD]

/* ****** ****** */

#define ATS_CURRENT_FRAME \
  ((ats_exception_frame_type*)the_ats_exception_stack)
#define ATS_CURRENT_FRAME_LVAL the_ats_exception_stack

#define ATS_ENTER_EXCEPTION_FRAME() \
  do { \
    ats_exception_frame_type *frame = \
      (ats_exception_frame_type*)alloca(sizeof(ats_exception_frame_type)); \
    frame->prev = ATS_CURRENT_FRAME ; \
    ATS_CURRENT_FRAME_LVAL = frame ; \
  } while (0) // end of [do]

#define ATS_LEAVE_EXCEPTION_FRAME() \
  do { ATS_CURRENT_FRAME_LVAL = ATS_CURRENT_FRAME->prev; } while (0)

#define ATS_RAISE(exn) \
  do { \
    if (ATS_CURRENT_FRAME == 0/*null*/) ats_uncaught_exception_handle(exn); \
    ATS_CURRENT_FRAME->exn = exn ; \
    siglongjmp(ATS_CURRENT_FRAME->env, 0) ; \
  } while (0) // end of [do]

/* ****** ****** */

/*
** WARNING: DO NOT USE THE FOLLOWING MACROS:
*/

#define ATS_TRYWITH_TRY(tmp_exn) \
do { \
ATS_ENTER_EXCEPTION_FRAME() ; \
tmp_exn = (ats_exn_ptr_type)(intptr_t)sigsetjmp(ATS_CURRENT_FRAME->env, 0) ; \
if ((intptr_t)tmp_exn == 0) { /* ... */

#define ATS_TRYWITH_WITH(tmp_exn) \
ATS_LEAVE_EXCEPTION_FRAME() ; \
} else { \
tmp_exn = ATS_CURRENT_FRAME->exn ; \
ATS_LEAVE_EXCEPTION_FRAME() ; /* exception handling */

#define ATS_TRYWITH_END() \
} \
} while(0)

/* end of WARNING */

/* ****** ****** */

static inline
ats_void_type ats_raise_exn // raising an exception
  (const ats_exn_ptr_type exn) { ATS_RAISE(exn) ; return ; }
/* end of [ats_raise_exn] */

/* ****** ****** */

/*
**
** function for generating new exception constructor tag
**
*/

extern int ats_exception_con_tag ;

static inline
int ats_exception_con_tag_new () {
  ats_exception_con_tag += 1 ; return (ats_exception_con_tag) ;
} /* end of [ats_exception_con_tag_new] */

/* functions for handling match failures */

extern
ats_void_type ats_caseof_failure(void) ;

extern
ats_void_type ats_funarg_match_failure(void) ;

/* ****** ****** */

#endif /* ATS_EXCEPTION_H */

/* end of [ats_exception.h] */
